File: algorithm/python/BOJ15649.py
# from https://www.acmicpc.net/problem/15649
import sys

N, M = map(int, sys.stdin.readline().split())

is_used = [False] * N
l = list(map(lambda x: x, range(1, N + 1)))
result = [0] * M


def permutation(cnt):
    if cnt == M:
        for num in result:
            print(num, end=" ")
        print()

        return

    for i in range(0, len(l)):
        if is_used[i]:
            continue
        is_used[i] = True
        result[cnt] = l[i]
        permutation(cnt + 1)
        is_used[i] = False


permutation(0)

File: algorithm/typescript/BOJ15649.ts
// from https://www.acmicpc.net/problem/15649

import fs = require('fs')

let str: string[] = fs.readFileSync('/dev/stdin').toString().split(' ');
let N: number = parseInt(str[0]);
let M: number = parseInt(str[1]);

let input = [];
for (let i = 1; i <= N; i++) {
	input.push(i);
}
let output = [];
let result = [];
let isSelected = Array(N).fill(false);

function permutation(cnt: number): void {
	if (cnt == M) {
		for (let i = 0; i < M; i++) {
			result.push(`${output[i]} `);
		}
		result.push(`\n`);
		return;
	}

	for (let i = 0; i < N; i++) {
		if (isSelected[i]) continue;
		isSelected[i] = true;
		output[cnt] = input[i];
		permutation(cnt + 1);
		isSelected[i] = false;
	}
}

permutation(0);
console.log(result.join(''));
